### ISSUE TITLE
Optimizing Custom Video Effects (IBasicVideoEffect)

### ISSUE THREAD

**Issue by ratishphilip** [Microsoft/Win2D#513]
I am trying to implement a custom video effect for my app by implementing the **IBasicVideoEffect** interface.

The ProcessFrame method mainly involves some transformation and cropping of the input frame to calculate the output frame.

Here is a sample code.

```C#
public void ProcessFrame(ProcessVideoFrameContext context)
{
    var inputSurface = context.InputFrame.Direct3DSurface;
    var outputSurface = context.OutputFrame.Direct3DSurface;

    using (var inputBitmap = CanvasBitmap.CreateFromDirect3D11Surface(canvasDevice, inputSurface))
    using (var renderTarget = CanvasRenderTarget.CreateFromDirect3D11Surface(canvasDevice, outputSurface))
    using (var ds = renderTarget.CreateDrawingSession())
    {
        // Some Transform2DEffect * CropEffect * Transform2DEffect on the inputBitmap

       // Render the final image
        ds.DrawImage(...);

    }
}
```

I ran this code to record videos on two mobiles - 
- Lumia 640 (resolution 1920x1080 at 30fps)
- Lumia 950 XL (resolution 3840 x 2160 at 30 fps)

Upon running the code I found both the phones stuttering, there was screen tearing on the preview video (during the recording), skipped frames in the resulting video.
The whole experience was not as smooth as I expected it to be.

I ran the code in Lumia 950 XL for resolution 1920x1080 at 30fps, still I noticed the mobile was struggling to keep up.

I added a stopwatch class to time the duration it takes for the ProcessFrame method. At 30 fps, the ProcessFrame method would be called every ~32 ms. But each process frame was taking around 25-70 ms depending on the transformation and cropping. I am on the latest Insider Build on each of the mobiles and am running the debug version of my app in them.

Is there a way I can optimize it further to achieve a buttery smooth performance?
Will implementing the custom video effect in C++/CX result in a better performance? 
What other factors influence the performance of the ProcessFrame method?

**Comment by shawnhar** [comment]
It doesn't look like you are doing anything particularly complicated inside the image effects, so this is most likely just limited by the size of the images being processed.  These are high resolution video feeds to be editing in realtime on a phone GPU!

If it is possible to replace your use of image effects with a single transformed DrawImage directly from the input bitmap, that /might/ be somewhat faster.

Otherwise the main thing that will improve performance is if you can reduce resolution of the video being processed.

> Will implementing the custom video effect in C++/CX result in a better performance?

Nope.  This kind of work is 100% GPU rather than CPU limited.

**Comment by ratishphilip** [comment]
What I am trying to achieve here is crop the input bitmap in such a way that the video is always in landscape (irrespective of the orientation of the phone). I am using the accelerometer to get the current orientation and calculating the angle at which the input image must be rotated and cropped.
All this calculation happens on a separate thread and the data is ready by the time ProcessFrame is called.
In the ProcessFrame the inputFrame is first rotated to balance the rotation of the phone, then cropped and then rotated back to original state.

> If it is possible to replace your use of image effects with a single transformed DrawImage directly from the input bitmap, that /might/ be somewhat faster.

Since it is a Transform2DEffect followed by CropEffect and then another Transform2D effect, how can I combine them? If it was just two Transform2D effects, I could have just obtained the final Matrix3x2. But I do I combine a CropEffect with Transform2D effect to get a single transformation?

**Comment by shawnhar** [comment]
I don't think you need to bother with image effects at all for that.  Just set CanvasDrawingSession.Transform for your rotation, then call DrawImage (specifying a source rect for cropping if necessary).

**Comment by ratishphilip** [comment]
Let me try that! Thank you.

**Comment by ratishphilip** [comment]
I tried setting the Transform for the DrawingSession but I am not getting the desired result. :(

This is what I want to achieve.

For a normal image i.e. when the phone is not tilted (shown in left) , the output image is shown in the right

![img1](https://cloud.githubusercontent.com/assets/7021835/25215722/1ab1aafe-2554-11e7-925b-d566a5880f06.png)

For a tilted image, the output image shall be a cropped and scaled up image
![img2](https://cloud.githubusercontent.com/assets/7021835/25215728/23381fe6-2554-11e7-9b20-a869f450c6f4.png)

When we specify a source rect for cropping in the DrawImage method, the source rect is specified in the inputBitmap's coordinate system.

So in my original method, when I applied a Transform2DEffect to rotate the inputBitmap, the inputBitmap's bounds also changes (shown in blue rectangle below). The CropEffect's SourceRectangle is calculated with respect to the new bounds.
![img3](https://cloud.githubusercontent.com/assets/7021835/25215741/2bd6e948-2554-11e7-9067-16ba88cd8864.png)

Whereas if the Transform is applied to the DrawingSession, the bounds of the inputBitmap does not change and thus the source rectangle used to crop the image will not be rotated. Instead it will be aligned with the rotated inputBitmap.

![img4](https://cloud.githubusercontent.com/assets/7021835/25215747/3440f8f8-2554-11e7-97b5-a607139452e0.png)
